R/rvar-rfun.R

```diff
@@ -0,0 +1,250 @@
+  }
+  formals(FUNV) <- formals(.f)
+  FUNV
+}
+
+#' Execute expressions of random variables
+#'
+#' Execute (nearly) arbitrary R expressions that may include [`rvar`]s,
+#' producing a new [`rvar`].
+#'
+#' @param expr A bare expression that can (optionally) contain [`rvar`]s. The expression
+#' supports [quasiquotation].
+#' @param ndraws When no [`rvar`]s are supplied in `expr`, this is the number
+#' of draws that will be used to construct new random variables. If `NULL`,
+#' `getOption("posterior.rvar_ndraws")` is used (default 4000). If `expr` contains
+#' [`rvar`]s, the number of draws in the provided [`rvar`]s is used instead of
+#' the value of this argument.
+#' @template args-rvar-dim
+#'
+#' @details This function evaluates `expr` possibly multiple times, once for each draw of
+#' the [`rvar`]s it contains, then returns a new [`rvar`] representing the output of those
+#' expressions. To identify [`rvar`]s, `rdo()` searches the calling environment for any variables
+#' named in `expr` for which [is_rvar()] evaluates to `TRUE`. If `expr` contains no [`rvar`]s,
+#' then it will be executed `ndraws` times and an [`rvar`] with that many draws returned.
+#'
+#' `rdo()` is not necessarily *fast* (in fact in some cases it may be very slow), but
+#' it has the advantage of allowing a nearly arbitrary R expression to be executed against [`rvar`]s
+#' simply by wrapping it with `rdo( ... )`. This makes it especially useful as a prototyping
+#' tool. If you create code with `rdo()` and it is unacceptably slow for your application,
+#' consider rewriting it using math operations directly on [`rvar`]s (which should be fast),
+#' using [rvar_r()], and/or using operations directly on the arrays that back the [`rvar`]s
+#' (via [draws_of()]).
+#'
+#' @return An [`rvar`].
+#'
+#' @examples
+#'
+#' mu <- rdo(rnorm(10, mean = 1:10, sd = 1))
+#' sigma <- rdo(rgamma(1, shape = 1, rate = 1))
+#' x <- rdo(rnorm(10, mu, sigma))
+#' x
+#'
+#' @family rfun
+#' @importFrom rlang eval_tidy quo_get_env enquo missing_arg quo_get_expr
+#' @export
+rdo <- function(expr, dim = NULL, ndraws = NULL) {
+  ndraws <- ndraws %||% getOption("posterior.rvar_ndraws", 4000)
+
+  # basic idea here is to find all the variables that are used in the expression
+  # and which are also random variables in the expression's environment, then
+  # build a function that executes the expression and takes those random
+  # variables as arguments, then vectorize that function using `rfun()` and execute it.
+  f_expr <- enquo(expr)
+  f_env <- quo_get_env(f_expr)
+
+  rvars_in_expr <- list()
+  f_alist <- alist()
+  for (name in all.vars(f_expr)) {
+    var <- get(name, f_env)
+    if (is_rvar(var)) {
+      rvars_in_expr[[name]] <- var
+      f_alist[[name]] <- missing_arg()
+    }
+  }
+
+  f_alist <- append(f_alist, quo_get_expr(f_expr))
+  f <- rfun(as.function(f_alist, envir = f_env), ndraws = ndraws)
+  result <- do_call(f, rvars_in_expr, enclos = f_env)
```

Comment:
I had to add an `enclos` argument to `do_call` so that I could replace `do.call` with `do_call` here. Let me know if you'd prefer a different solution.

Reply:
Looks good to me. I have had way to many problems with do.call in brms, especially if two or more do.calls happend to be nested in one-another, so I wanted to make sure, we do not run into the same trouble here.